src/common/dialog.ts

```diff
@@ -443,171 +251,426 @@ class Dialog extends Panel {
+     */
+    createHeader(title: string): Widget {
+      let header = new Widget();
+      header.addClass('jp-Dialog-header');
+      let titleNode = document.createElement('span');
+      titleNode.textContent = title;
+      titleNode.className = 'jp-Dialog-title';
+      header.node.appendChild(titleNode);
+      return header;
+    }
 
-  private _buttonNodes: HTMLButtonElement[];
-  private _buttons: IButtonItem[];
-  private _original: HTMLElement;
-  private _first: HTMLElement;
-  private _primary: HTMLElement;
-}
+    /**
+     * Create the body of the dialog.
+     *
+     * @param value - The input value for the body.
+     *
+     * @returns A widget for the body.
+     */
+    createBody(value: BodyType): Widget {
+      let body: Widget;
+      if (typeof value === 'string') {
+        body = new Widget({ node: document.createElement('span') });
+        body.node.textContent = value;
+      } else if (value instanceof Widget) {
+        body = value;
+      } else {
+        body = new Widget({ node: value });
+      }
+      body.addClass('jp-Dialog-body');
+      Styling.styleNode(body.node);
+      return body;
+    }
 
+    /**
+     * Create the button nodes of the dialog.
+     *
+     * @param buttons - The buttons data.
+     *
+     * @returns An array of elements for the buttons.
+     */
+    createButtonNodes(buttons: ReadonlyArray<IButton>): ReadonlyArray<HTMLElement> {
+      let nodes = new Array<HTMLElement>(buttons.length);
+      each(buttons, (button, i) => {
+        nodes[i] = VirtualDOM.realize(this.renderButtonNode(button));
+      });
+      return nodes;
+    }
 
-/**
- * Create a dialog body widget from a non-widget input.
- */
-function createDialogBody(body: HTMLElement | string): Widget {
-  let child: HTMLElement;
-  if (typeof body === 'string') {
-    child = document.createElement('span');
-    child.innerHTML = body as string;
-  } else if (body) {
-    child = body as HTMLElement;
-    switch (child.tagName) {
-    case 'INPUT':
-      child = wrapInput(child as HTMLInputElement);
-      break;
-    case 'SELECT':
-      child = wrapSelect(child as HTMLSelectElement);
-      break;
-    default:
-      child = styleElements(child);
-      break;
+    /**
+     * Create the footer of the dialog.
+     *
+     * @param buttons - The buttons to add to the footer.
+     *
+     * @returns A widget for the footer.
+     */
+    createFooter(buttons: ReadonlyArray<HTMLElement>): Widget {
+      let footer = new Widget();
+      footer.addClass('jp-Dialog-footer');
+      each(buttons, button => {
+        footer.node.appendChild(button);
+      });
+      Styling.styleNode(footer.node);
+      return footer;
     }
-  }
-  child.classList.add(BODY_CONTENT_CLASS);
-  return new Widget({node: child});
-}
 
+    /**
+     * Render a button node for the dialog.
+     *
+     * @param data - The data to use for rendering the button.
+     *
+     * @returns A virtual element representing the button.
```

Comment:
I'm not sure I understand the benefit of using virtual DOM here if it gets immediately `realize`d.